File: CodingSamples/Java/Inheritance/InterfaceTest2/Program1.java
import taxation.*;

class Program {

	//nested member class is defined inside of another class with static modifier
	static class Auditor {
		
		public Auditor() {
			System.out.println("Acquiring resources...");
		}

		public double audit(TaxPayer person, String id) {
			if(id.length() < 4)
				throw new IllegalArgumentException("Invalid ID");
			return person.incomeTax() + 500;
		}

		public void close() {
			System.out.println("Releasing resources...");
		}
	}

	private static void doAudit(String name, int count) {
		TaxPayer emp = name.equals("jack") ? new Engineer(count) : new Programmer(count);
		var a = new Auditor(); //type inference from initializer (var = Auditor)
		try{
			System.out.printf("Total Payment: %.2f%n", a.audit(emp, name));
		}finally{ //finally block always executes before control leaves try block
			a.close();
		}
	}

	public static void main(String[] args) {
		try{
			String m = args[0].toLowerCase();
			int n = Integer.parseInt(args[1]);
			doAudit(m, n);

		}catch(Exception e){
			System.out.printf("Error: %s%n", e);
		}
	}
}

File: CodingSamples/DotNet/Inheritance/InterfaceTest2/DemoApp/Program1.cs
using Taxation;

try
{
    string m = args[0].ToLower();
    int n = int.Parse(args[1]);
    DoAudit(m, n);
}
catch(Exception ex)
{
    Console.WriteLine(ex.Message);
}

void DoAudit(string name, int count)
{
    //auto-boxing for converting struct (value) type to interface (reference) type
    ITaxPayer emp = name == "jack" ? new Engineer(count) : new Programmer(count);
    var a = new Auditor();
    try
    {
        Console.WriteLine("Total Payment: {0:0.00}", a.Audit(emp, name));
    }
    finally
    {
        a.Dispose();
    }
}

class Auditor
{
    public Auditor()
    {
        Console.WriteLine("Acquiring resources...");
    }

    public double Audit(ITaxPayer person, string id)
    {
        if(id.Length < 4)
            throw new ArgumentException("Invalid ID");
        return person.IncomeTax() + 500;
    }

    public void Dispose()
    {
        Console.WriteLine("Releasing resources...");
    }
}

